### ISSUE TITLE
[FEATURE] Anyone developing rules to satisfy ADA requirements in HTML templates?

### ISSUE THREAD

**Issue by bulldoguk** [Riverside-Healthcare/djLint#718]
## Feature Request
The custom rules feature works like a charm, but before I dig into all the regex, etc needed to check for ARIA tags, etc to satisfy ADA compliance - has anyone / is anyone already working on this so we can combine forces?

**Comment by christopherpickering** [comment]
Hey @bulldoguk thanks for brining this one up, it would be great to have a set of rules for these. Is there a more readable doc than the government one on what tag/attribute combinations are required? https://www.ada.gov/law-and-regs/design-standards/

**Comment by juleskuehn** [comment]
The presence or absence of ARIA tags won't determine Section 508 compliance.

**Comment by monosans** [comment]
Picking this up, with @juleskuehn's caveat up front, because it shapes what djLint can honestly offer: a linter cannot determine Section 508 or ADA compliance. Most of WCAG is contextual, and whether an alt text is accurate, or a contrast ratio passes, or a page reads in a sensible order, is not visible in a template. What a static checker can do is catch specific failures that are decidable from the markup alone.

Already there before v1.45.0:

| Rule | Check |
| ---- | ----- |
| H005 | `<html>` has a non-empty `lang` |
| H013 | `<img>` has an `alt` |
| H016 | the document has a `<title>` |
| H025 | tags are closed, so structure is not left to the parser to guess |
| H036 | `<br>` is not used for layout |
| H042 | a `<label for>` matches an `id` in the file |

Nine more arrived in v1.45.0, all on by default:

| Rule | Check |
| ---- | ----- |
| H043 | `<button>` has a `type` |
| H044 | a header row does not mix `th` and `td` |
| H045 | `<iframe>` has an accessible name |
| H046 | `tabindex` is not positive |
| H047 | `aria-hidden="true"` is not set on an element that takes focus |
| H048 | an `aria-` attribute is one ARIA defines |
| H049 | the viewport does not stop the page being zoomed |
| H051 | a `role` is one ARIA defines for markup |
| H052 | `<meta http-equiv="refresh">` does not reload or redirect on a timer |

`H036` is on by default from v1.45.0 too, where it used to be opt-in. It now reports only the two uses of `<br>` the html specification rules out, a run of two or more and a break against the inside edge of a block element, so a break that is part of the content is left alone.

A positive `tabindex` rearranges the tab order for the whole page, `aria-hidden` on a focusable element strands a keyboard user on a control nothing will announce, and a misspelled `aria-` attribute does nothing at all while looking as though it did. None of the three is visible in review, which is what makes them worth a linter.

`H045` as an example:

```html
<iframe src="/report/"></iframe>
<iframe src="/report/" title="Quarterly report"></iframe>
```

The first is reported, the second is not. A screen reader announces a frame by its name, and without one it reads the url or says nothing, so there is no way to tell what is embedded before entering it. `title`, `aria-label` and `aria-labelledby` all satisfy it, and a name written by a template tag counts, so a frame titled per page is not reported.

All of them stay quiet where a template makes the answer unknowable. A `tabindex`, `role` or aria value written by a template tag is left alone, as is a framework binding such as `:role` or `[attr.aria-label]`, since djLint cannot see what either renders to.

The same invitation stands for anything not on those lists. The useful contribution is not a rule set called ADA but a single check that is decidable from markup and close enough to zero false positives to be on by default. Open an issue per check with a case that should fire and a case that should not, and it can go in.
